cpp/dynamic_community_detection_module/algorithm/dynamic_community_detection.hpp

```diff
@@ -0,0 +1,241 @@
+  /// exponent used in the inflation step
+  double exponent = 4;
+  /// lowest probability at the cutoff step
+  double min_value = 0.1;
+
+  /// weight property name
+  std::string weight_property = "weight";
+  /// default self-loop weight
+  double w_selfloop = 1.0;
+
+  /// maximum number of iterations
+  std::uint64_t max_iterations = 100;
+  /// maximum number of updates for any node
+  std::uint64_t max_updates = 5;
+  /* #endregion */
+
+  /* #region structures */
+  /// reference to current graph
+  std::unique_ptr<mg_graph::Graph<>> graph;
+
+  /// map containing each node’s community label probabilities
+  std::unordered_map<std::uint64_t, std::unordered_map<std::uint64_t, double>>
+      label_Ps;
+  /// sum of weights of each node’s in-edges
+  std::unordered_map<std::uint64_t, double> sum_w;
+  /// how many times each node has been updated
+  std::unordered_map<std::uint64_t, std::uint64_t> times_updated;
+
+  /// flag whether community labels have already been calculated
+  bool calculated = false;
+  /* #endregion */
+
+  /* #region helper_methods */
+  ///@brief Returns a vector of all graph nodes’ Memgraph IDs.
+  ///
+  ///@return -- vector of all graph nodes’ Memgraph IDs
+  std::vector<std::uint64_t> NodesMemgraphIDs();
+
+  ///@brief Returns in-neighbors’ Memgraph IDs for a given node’s Memgraph ID.
+  /// If the graph is undirected, all neighbors are included.
+  ///
+  ///@param node_id -- given node’s Memgraph ID
+  ///
+  ///@return -- vector of neighbors’ Memgraph IDs
+  std::unordered_set<std::uint64_t> InNeighborsMemgraphIDs(
+      const std::uint64_t node_id);
+
+  ///@brief Initializes internal data structures.
+  void SetStructures(const std::uint64_t node_id);
+
+  ///@brief Removes deleted nodes’ entries from internal data structures.
+  ///
+  ///@param deleted_nodes_ids -- set of deleted nodes’ IDs
+  void RemoveDeletedNodes(const std::unordered_set<std::uint64_t>& deleted_nodes_ids);
+
+  ///@brief Retrieves the total weight of edges between two nodes.
+  /// With undirected graphs it considers all edges between the given node pair.
+  /// On a directed graph, it checks if the source and target nodes match.
+  ///
+  ///@param from_node_id -- source node’s Memgraph ID
+  ///@param to_node_id -- target node’s Memgraph ID
+  ///
+  ///@return -- total weight of edges between from_node_id and to_node_id
+  double GetTotalWeightBetween(const std::uint64_t from_node_id,
+                               const std::uint64_t to_node_id);
+
+  ///@brief Retrieves given edge’s weight.
+  double GetWeight(const std::uint64_t edge_id);
+
+  ///@brief Resets each node’s number of updates.
+  void ResetTimesUpdated();
+
+  /// Returns given node’s current community label.
+  /// If no label was found (e.g. with the min_value set too high),
+  /// the assigned label number is -1.
+  ///
+  /// @return -- given node’s current community label
+  std::int64_t NodeLabel(const std::uint64_t node_id);
```

Comment:
You can omit `const` on a declaration when you pass by value and add it only on its definition (that is the source file). That way,  API  users don't get confused (from their point of view, they don't care if the function modifies its arguments because of the by value semantics)

Reply:
If we take into account that this API is used by this query module only, should these still be omitted?

Reply:
Yes, the query module shall only care that the value does not change (since we pass by value) but the implementation of NodeLabel (in the source file) cares if the that argument is modified on its body. That way you get a clean separation. I hope that helps :)

Reply:
Thanks! I’ll be making this change in line with your comment and Google’s style guide which we try to follow at MAGE.